HADOOP-18073. Client config, and list, put, copy and head operations.

ahmarsuhail opened the pull request:
### Description of PR

This PR updates client config, and the list, put, copy and head object operations. 

When reviewing, it's probably easier to separate by commits.

- [Changes for client and config and update list operations](https://github.com/ahmarsuhail/hadoop/pull/3/commits/77c5cf04dc148fca7b1c5874be69b1a2cd47ed21) - These have already been reviewed by @dannycjones [here](https://github.com/apache/hadoop/pull/4706). That PR has been closed without merging, and will be replaced by this one. 
- [Changes for Copy, Put & Head](https://github.com/ahmarsuhail/hadoop/pull/3/commits/77a7051473607663ad22a25b0bfc1e4bee54b3dc) : Changes made are explained in [Overview of changes for GetObjectMetadata, PutObject & CopyObject.pdf](https://github.com/ahmarsuhail/hadoop/files/9663205/Overview.of.changes.for.GetObjectMetadata.PutObject.CopyObject.pdf).

### Still TODO

There are some issues that I'm still looking into:

- Something goes wrong when you add in the Transfer Manager dependency. In `AWSClienConfig`, `org.apache.http.client.utils.URIBuilder;` is no longer available. And get a `class file for org.apache.commons.logging.Log not found` in `S3ADelegationTokens`. Also when I try to use progress listener with the TM, I get a log4j error. Will look into why this is happening and update. 
 - Previously a call to getObjectMetadata for a base path, ie with an empty key would return some metadata. (bucket region, content type). headObject() fails without a key.
 - The TM does not return `copyObjectResult` which has the version ID, which causes an issue in the ChangeTracker.

### How was this patch tested?

Tested in eu-west-1 by running `mvn -Dparallel-tests -DtestsThreadCount=16 clean verify`

The following tests are failing: 

Test Failing | Reason
-- | --
ITestS3AAWSCredentialsProvider.testBadCredentialsConstructor | Fails because SDKV2 throws SdkException, will work once errors are handled and translated properly.
ITestAuditAccessChecks.testDirAccessDenied | Fails because list requests are not wired up to auditor yet
ITestAuditAccessChecks.testFileAccessDenied | Fails because list requests are not wired up to auditor yet
ITestAuditAccessChecks.testMissingPathAccessFNFE | Fails because list requests are not wired up to auditor yet
ITestAuditManager.testRequestHandlerBinding | Fails because list requests are not wired up to auditor yet
ITestAWSStatisticCollection.testCommonCrawlStatistics | Fails as not tied up to auditor yet, so does not update stats when getObjectMetadata is called
ITestAWSStatisticCollection.testLandsatStatistics | Fails as not tied up to auditor yet, so does not update stats when getObjectMetadata is called
ITestMarkerTool.testRunWrongBucket | Throws NoSuchBucketException, which is not yet translated
ITestS3AConfiguration.testAutomaticProxyPortSelection | Fails because SDKV2 throws SdkException, will work once errors are handled and translated properly.
ITestS3AConfiguration.testProxyConnection | Fails because SDKV2 throws SdkException, will work once errors are handled and translated properly.
ITestS3AConfiguration.shouldBeAbleToSwitchOnS3PathStyleAccessViaConfigProperty | Fails due to errors not being translated It throws Software.S3Exception: The bucket you are attempting to access must be addressed using the specified endpoint. will be fixed during error translation
ITestS3AStorageClass.testCreateAndCopyObjectWithStorageClassGlacier | Copy works ok, but fails because it expects AccessDeniedException when it tries to get the object from glaciers, and errors are not translated yet. Will be updated during error translation work.
ITestS3ATemporaryCredentials.testInvalidSTSBinding, testSTS | Fails because it throws a 400, which is translated to AWSBadRequestException. Will be fixed during error translation.
ITestSessionDelegationInFileystem.testDelegatedFileSystem, ITestXAttrCost.testXAttrRoot | Fails because tries to do headObject() with empty key which now fails, need to look at how to fix. Note: ITestSessionDelegationInFileystem.testDelegatedFileSystem does not fail as i've temporarily removed the headobject call from it

ahmarsuhail committed:
configures s3 client, updates list operation

ahmarsuhail committed:
updates getObjectMetadata, putObject & copyObject operations

passaro commented on hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/impl/ChangeDetectionPolicy.java:
```diff
@@ -362,14 +363,14 @@ public void applyRevisionConstraint(CopyObjectRequest request,
       if (revisionId != null) {
         LOG.debug("Restricting copy request to etag {}", revisionId);
-        request.withMatchingETagConstraint(revisionId);
+        request.toBuilder().copySourceIfMatch(revisionId);
```
I suspect that the call to `toBuilder()` creates a new instance rather than modify the original `request`.

passaro committed:
Upgrade GetObject to use SDK v2.

Key change: `getObject` now returns a `ResponseInputStream<GetObjectResponse>` rather than a `S3Object`. This makes it simpler to handle the input stream lifetime in various classes such as `S3AInputStream`, `S3ARemoteObject`, or `SDKStreamDrainer`.

ahmarsuhail committed:
fixes failing unit tests

ahmarsuhail committed:
updates requestFactory to return builder

ahmarsuhail committed:
renames factory methods to builder

ahmarsuhail committed:
updates content range logic

ahmarsuhail committed:
Merge pull request #4 from passaro/HADOOP-18073-getObject

Upgrade GetObject to use SDK V2.
